django/db/models/expressions.py

```diff
@@ -305,7 +308,7 @@ def prepare_database_save(self, field):
         return self
 
     def get_group_by_cols(self):
-        if not self.contains_aggregate:
+        if not self.contains_aggregate and self.allowed_in_group_by:
```

Comment:
Commit 0c7633178fa9410f102e4708cef979b873bccb76 broke tests where a case expression referenced an aggregate. The whole case SQL was added to the group by clause, I've added an ``allowed_in_group_by`` attribute on ``ExpressionNode`` which subclasses (``When`` and ``Case``) can set to ``False`` to force processing of  child expressions instead of returning ``self``. This works but I wonder if I am using ``contains_aggregate`` correctly.